packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -245,41 +260,293 @@ class FilteringTextInputFormatter extends TextInputFormatter {
-        );
-      },
-    );
+    final TextEditingValueAccumulator accumulator = TextEditingValueAccumulator._(newValue);
+
+    final Iterable<Match> matches = filterPattern.allMatches(newValue.text);
+    Match? previousMatch;
+    for (final Match match in matches) {
+      final int allowStart;
+      final int allowEnd;
+      final int replaceStart;
+      final int replaceEnd;
+
+      assert(match.end > match.start);
+
+      // Compute the non-match region between this `Match` and the previous
+      // `Match`. Depending on the value of `allow`, either the match region or
+      // the non-match region is the banned pattern.
+      if (allow) {
+        // The non-match region is the banned pattern.
+        allowStart = match.start;
+        allowEnd = match.end;
+        replaceStart = previousMatch?.end ?? 0;
+        replaceEnd = match.start;
+
+        // Apply the replacingStrategy first since the non-match region is
+        // banned.
+        replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+        accumulator.stringBuffer.write(newValue.text.substring(allowStart, allowEnd));
+      } else {
+        // When `allow` is false the match region is the banned pattern.
+        allowStart = previousMatch?.end ?? 0;
+        allowEnd = match.start;
+        replaceStart = match.start;
+        replaceEnd = match.end;
+
+        // Write the allowed word to the buffer first.
+        accumulator.stringBuffer.write(newValue.text.substring(allowStart, allowEnd));
+        replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+      }
+
+      previousMatch = match;
+    }
+
+    // Handle the last non-match region betwee the last match region and the end
+    // of the text.
+    final int replaceStart = previousMatch?.end ?? 0;
+    final int replaceEnd = newValue.text.length;
+
+    if (allow) {
+      replacingStrategy(accumulator, replaceStart, replaceEnd, replacementString);
+    } else {
+      accumulator.stringBuffer.write(newValue.text.substring(replaceStart));
+    }
+
+    return accumulator._finalize();
   }
 
   /// A [TextInputFormatter] that forces input to be a single line.
   static final TextInputFormatter singleLineFormatter = FilteringTextInputFormatter.deny('\n');
 
   /// A [TextInputFormatter] that takes in digits `[0-9]` only.
   static final TextInputFormatter digitsOnly = FilteringTextInputFormatter.allow(RegExp(r'[0-9]'));
+
+  /// A [FilteringFormatterReplacingStrategy] that attempts to preserve the
+  /// pre-format selection and composing region as much as possible.
+  ///
+  /// If the start or the end of the selection (or the composing region) is
+  /// strictly inside the banned pattern described by `regionStart` and
+  /// `regionEnd` (meaning, `regionStart` < index < `regionEnd`), the index will
+  /// be placed at the end of the replacement string.
+  static void preserveSelectionAndComposingRegion(
```

Comment:
The discoverability for these 2 strategies isn't great. Maybe they should be moved to a separate class.

Reply:
There should be a see also section on the doc of replacingStrategy

Reply:
This feel weird as an API. we make customizing selection and composing flexible in an opinionated class that is not flexible about customizing the replacestring.

Reply:
also the selection and composing usually depends on the entire string instead of individual region of the text. I was wondering whether we can call this strategy once with all the formatting information instead. @goderbauer may have better idea

Reply:
Could you elaborate on "opinionated class"?

Reply:
Oh you are referring to "not flexible about customizing the replacestring"? I think in a strategy implementation you can put whatever text you want in the `StringBuffer`, instead of the `replacementString`.

Reply:
Actually I think I can make the API slightly safer by asserting after each application of `replacingStrategy` that the new index is not OOB. Lemme do that.